drivers/adc/adc_ite_it8xxx2.c

```diff
@@ -0,0 +1,337 @@
+			     uint8_t active_channels)
+{
+	size_t needed_buffer_size;
+
+	needed_buffer_size = active_channels * sizeof(uint16_t);
+
+	if (sequence->options) {
+		needed_buffer_size *= (1 + sequence->options->extra_samplings);
+	}
+
+	if (sequence->buffer_size < needed_buffer_size) {
+		LOG_ERR("Provided buffer is too small (%u/%u)",
+				sequence->buffer_size, needed_buffer_size);
+		return -ENOMEM;
+	}
+
+	return 0;
+}
+
+static int adc_it8xxx2_start_read(const struct device *dev,
+			      const struct adc_sequence *sequence)
+{
+	struct adc_it8xxx2_data *data = DEV_DATA(dev);
+	int err;
+
+	data->resolution = sequence->resolution;
+	LOG_DBG("Configure resolution=%d", sequence->resolution);
+
+	if (!sequence->channels ||
+		sequence->channels & ~BIT_MASK(IT8XXX2_CHANNEL_COUNT)) {
+		LOG_ERR("Invalid selection of channels");
+		return -EINVAL;
+	}
+
+	err = check_buffer_size(sequence, 1);
+	if (err) {
+		return err;
+	}
+
+	data->buffer = sequence->buffer;
+
+	adc_context_start_read(&data->ctx, sequence);
+
+	return adc_context_wait_for_completion(&data->ctx);
+}
+
+static void adc_context_start_sampling(struct adc_context *ctx)
+{
+	struct adc_it8xxx2_data *data =
+		CONTAINER_OF(ctx, struct adc_it8xxx2_data, ctx);
+
+	data->repeat_buffer = data->buffer;
+
+	/* bit 0 : adc module enable */
+	IT83XX_ADC_ADCCFG |= 0x01;
+}
+
+static int adc_it8xxx2_read(const struct device *dev,
+			const struct adc_sequence *sequence)
+{
+	struct adc_it8xxx2_data *data = DEV_DATA(dev);
+	int err;
+
+	adc_context_lock(&data->ctx, false, NULL);
+	err = adc_it8xxx2_start_read(dev, sequence);
+	adc_context_release(&data->ctx, err);
+
+	return err;
+}
+
+static void adc_context_update_buffer_pointer(struct adc_context *ctx,
+					      bool repeat_sampling)
+{
+	struct adc_it8xxx2_data *data =
+		CONTAINER_OF(ctx, struct adc_it8xxx2_data, ctx);
+
+	if (repeat_sampling) {
+		data->buffer = data->repeat_buffer;
+	}
+}
+
+static int adc_data_valid(enum chip_adc_channel adc_ch)
+{
+	return IT83XX_ADC_ADCDVSTS & BIT(adc_ch);
+}
+
+static void adc_it8xxx2_get_sample(const struct device *dev)
+{
+	struct adc_it8xxx2_data *data = DEV_DATA(dev);
+	uint8_t valid = 0;
+
+	if (adc_data_valid(data->adc_ch)) {
+
+		*data->buffer = ((*adc_ctrl_regs[data->adc_ch].adc_datm << 8) +
+						*adc_ctrl_regs[data->adc_ch].adc_datl) *
+						data->resolution;
+
+		data->buffer++;
+
+		/* W/C data valid flag */
+		IT83XX_ADC_ADCDVSTS = BIT(data->adc_ch);
+		valid = 1;
+	}
+
+	if (!valid) {
+		printk("ADC failed to read!!! (regs=%x, ch=%d)",
+			IT83XX_ADC_ADCDVSTS,
+			data->adc_ch);
+	}
+	/* disable adc interrupt */
+	adc_disable_channel(dev, data->adc_ch);
+}
+
+static void adc_it8xxx2_isr(const void *arg)
+{
+	struct device *dev = (struct device *)arg;
+	struct adc_it8xxx2_data *data = DEV_DATA(dev);
+
+	adc_it8xxx2_get_sample(dev);
+
+	adc_context_on_sampling_done(&data->ctx, dev);
+
+	LOG_DBG("ADC ISR triggered.");
+}
+
+struct adc_driver_api api_it8xxx2_driver_api = {
```

Comment:
Is there a reason why this is not `static const`?

Reply:
Done